File: python/mirage/observe/disk_store.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import os
import shutil

import aiofiles
import aiofiles.os


class DiskObserverStore:
    """ObserverStore backed by a directory of JSONL files.

    Args:
        root (str): Directory holding the recorder's files; created on
            first append.
    """

    def __init__(self, root: str) -> None:
        self._root = root

    def _abs(self, path: str) -> str:
        return os.path.join(self._root, path.lstrip("/"))

    async def append(self, path: str, data: bytes) -> None:
        """Append bytes to the file at path, creating parents.

        Args:
            path (str): File key.
            data (bytes): Bytes to append.
        """
        abs_path = self._abs(path)
        await aiofiles.os.makedirs(os.path.dirname(abs_path), exist_ok=True)
        async with aiofiles.open(abs_path, "ab") as f:
            await f.write(data)

    async def write(self, path: str, data: bytes) -> None:
        """Overwrite the file at path, creating parents.

        Args:
            path (str): File key.
            data (bytes): Full new content.
        """
        abs_path = self._abs(path)
        await aiofiles.os.makedirs(os.path.dirname(abs_path), exist_ok=True)
        async with aiofiles.open(abs_path, "wb") as f:
            await f.write(data)

    async def read_all(self) -> dict[str, bytes]:
        """Read every stored file under root.

        Returns:
            dict[str, bytes]: Mapping of file key to content.
        """
        return await self._read_files(None)

    async def read_matching(self, suffix: str) -> dict[str, bytes]:
        """Read only the files whose key ends with suffix.

        Args:
            suffix (str): File-key suffix.

        Returns:
            dict[str, bytes]: Mapping of matching key to content.
        """
        return await self._read_files(suffix)

    async def clear(self) -> None:
        """Delete every stored file (snapshot-restore rewind)."""
        if os.path.isdir(self._root):
            shutil.rmtree(self._root)

    async def _read_files(self, suffix: str | None) -> dict[str, bytes]:
        out: dict[str, bytes] = {}
        if not os.path.isdir(self._root):
            return out
        for dirpath, _dirs, files in os.walk(self._root):
            for name in files:
                abs_path = os.path.join(dirpath, name)
                rel = "/" + os.path.relpath(abs_path, self._root)
                if suffix is not None and not rel.endswith(suffix):
                    continue
                async with aiofiles.open(abs_path, "rb") as f:
                    out[rel] = await f.read()
        return out

File: typescript/packages/node/src/observe/disk_store.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { appendFile, mkdir, readdir, readFile, rmdir, unlink, writeFile } from 'node:fs/promises'
import path from 'node:path'
import type { ObserverStore } from '@struktoai/mirage-core'

/**
 * ObserverStore backed by a directory of JSONL files. Created lazily on
 * first append. Mirrors the Python DiskObserverStore.
 */
export class DiskObserverStore implements ObserverStore {
  private readonly root: string

  constructor(root: string) {
    this.root = root
  }

  private abs(key: string): string {
    return path.join(this.root, key.replace(/^\/+/, ''))
  }

  async append(key: string, data: Uint8Array): Promise<void> {
    const absPath = this.abs(key)
    await mkdir(path.dirname(absPath), { recursive: true })
    await appendFile(absPath, data)
  }

  async write(key: string, data: Uint8Array): Promise<void> {
    const absPath = this.abs(key)
    await mkdir(path.dirname(absPath), { recursive: true })
    await writeFile(absPath, data)
  }

  readAll(): Promise<Map<string, Uint8Array>> {
    return this.readMatching('')
  }

  async readMatching(suffix: string): Promise<Map<string, Uint8Array>> {
    const out = new Map<string, Uint8Array>()
    const { files } = await this.walk()
    for (const absPath of files) {
      const rel = '/' + path.relative(this.root, absPath)
      if (!rel.endsWith(suffix)) continue
      out.set(rel, new Uint8Array(await readFile(absPath)))
    }
    return out
  }

  async clear(): Promise<void> {
    const { files, dirs } = await this.walk()
    for (const absPath of files) await unlink(absPath)
    for (const d of [...dirs].reverse()) await rmdir(d)
  }

  close(): Promise<void> {
    return Promise.resolve()
  }

  private async walk(): Promise<{ files: string[]; dirs: string[] }> {
    const files: string[] = []
    const dirs: string[] = []
    const stack: string[] = [this.root]
    while (stack.length > 0) {
      const d = stack.pop()
      if (d === undefined) break
      let entries
      try {
        entries = await readdir(d, { withFileTypes: true })
      } catch {
        continue
      }
      dirs.push(d)
      for (const entry of entries) {
        const full = path.join(d, entry.name)
        if (entry.isDirectory()) stack.push(full)
        else files.push(full)
      }
    }
    return { files, dirs }
  }
}
